### PR TITLE
Add option to pass file with dependencies to fetch

### PR DESCRIPTION
**Pull request by jkciesluk**
Connected to https://github.com/coursier/coursier/issues/2443

Allows to pass file based parameters to `resolve` and `fetch`.
Usage: 
`cs resolve --dependency-file <file>` or  with multiple files `cs resolve --dependency-file <file1>  --dependency-file <file2> `

### CODE REVIEW

File: modules/cli/src/main/scala/coursier/cli/params/DependencyParams.scala [hunk 1]
```diff
@@ -57,6 +58,23 @@ object DependencyParams {
...
 
+    val fromFilesDependenciesV: ValidatedNel[String, Seq[String]] =
+      if (options.dependencyFile.isEmpty)
+        Validated.validNel(Seq.empty)
+      else {
+        val files = options.dependencyFile
+        val fromFileDependencies = files.flatMap { file =>
+          val source = Source.fromFile(file)
+          val lines =
+            try source.mkString.split("\n").filter(_.nonEmpty)
```
Comment (jkciesluk) [inline modules/cli/src/main/scala/coursier/cli/params/DependencyParams.scala:69]: Maybe we should also allow separating dependencies with whitespace or e.g. `,` ?

**Review by julienrf — CHANGES_REQUESTED** [review]
Thank you for the contribution @jkciesluk!

I think the general approach looks fine.

Would it be possible to add more tests? For instance, what happens if a user passes a path to a non-existing file? Do we get a crash or a nice error message?

Could you also mention the --dependencyFile option in the documentation and give an example? https://get-coursier.io/docs/cli-fetch

**Comment by jkciesluk** [comment]
> Would it be possible to add more tests? For instance, what happens if a user passes a path to a non-existing file? Do we get a crash or a nice error message?

It ends with an exception, like it did for option `--local-exclude-file`, but maybe we can change it for both to give a nice error message. 


> Could you also mention the --dependencyFile option in the documentation and give an example? https://get-coursier.io/docs/cli-fetch

Sure, I will add it

**Comment by julienrf** [comment]
> It ends with an exception, like it did for option `--local-exclude-file`, but maybe we can change it for both to give a nice error message.

It would be great if that was covered by the tests.

**Comment by jkciesluk** [comment]
While working on coursier I can't compile my code because of this exception
```
Compiling /Users/jciesluk/coursier/coursier/build.sc
[1/300] coursier.jvm[2.13.6].publishVersion
fatal: no tag exactly matches 'bd1374911ffe5458b05a117b75b2e22bc2f53e1b'
fatal: No names found, cannot describe anything.
1 targets failed
coursier.jvm[2.13.6].publishVersion os.SubprocessException: CommandResult 128

    os.proc.call(ProcessOps.scala:85)
    ammonite.$file.project.modules.shared.latestTaggedVersion(shared.sc:20)
    ammonite.$file.project.modules.shared.latestTaggedVersion$(shared.sc:19)
    ammonite.$file.project.modules.shared$.latestTaggedVersion$lzycompute(shared.sc:48)
    ammonite.$file.project.modules.shared$.latestTaggedVersion(shared.sc:48)
    ammonite.$file.project.modules.shared.$anonfun$buildVersion$2(shared.sc:32)
    scala.Option.getOrElse(Option.scala:201)
    ammonite.$file.project.modules.shared.buildVersion(shared.sc:30)
    ammonite.$file.project.modules.shared.buildVersion$(shared.sc:22)
    ammonite.$file.project.modules.shared$.buildVersion$lzycompute(shared.sc:48)
    ammonite.$file.project.modules.shared$.buildVersion(shared.sc:48)
    ammonite.$file.project.modules.shared$CoursierPublishModule.$anonfun$publishVersion$2(shared.sc:52)
    mill.define.Task$TraverseCtx.evaluate(Task.scala:380)
```

 I've managed to fix this by removing `--exact-match` from https://github.com/coursier/coursier/blob/de7ac23f7c2b896e6665658dab1437099a0a3fc0/project/modules/shared.sc#L25
 
 But now, I can't do `$ ./mill -i doc.generate` because of 
 ```[300/300] doc.generate
fatal: No names found, cannot describe anything.
1 targets failed
doc.generate os.SubprocessException: CommandResult 128

    os.proc.call(ProcessOps.scala:85)
    ammonite.$file.project.modules.shared.latestTaggedVersion(shared.sc:20)
    ammonite.$file.project.modules.shared.latestTaggedVersion$(shared.sc:19)
    ammonite.$file.project.modules.shared$.latestTaggedVersion$lzycompute(shared.sc:48)
    ammonite.$file.project.modules.shared$.latestTaggedVersion(shared.sc:48)
    ammonite.$file.project.modules.doc0$Doc.$anonfun$generate$2(doc0.sc:50)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
```

Any idea how to fix this or why I have to remove this flag from `shared.sc`?

File: doc/docs/cli-fetch.md [hunk 2]
```diff
@@ -59,6 +59,23 @@ $ cs fetch --javadoc com.lihaoyi:ammonite_2.12.8:1.6.0
...
+$ cs fetch --dependency-file file_1 --dependency-file file_2
+/Users/jciesluk/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/circe/circe-generic_2.13/0.12.3/circe-generic_2.13-0.12.3.jar
+/Users/jciesluk/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite_2.12.8/1.6.0/ammonite_2.12.8-1.6.0.jar
+…
+```
+With `file_1` content:
+```
+io.circe::circe-generic:0.12.3
+com.lihaoyi:ammonite_2.12.8:1.6.0
+```
```
Comment (julienrf) [inline doc/docs/cli-fetch.md]: In this example, the fact that there is no `file_2` is not an error?

Comment (jkciesluk) [reply doc/docs/cli-fetch.md]: I wanted to use two files in command to show that user can do that, but I gave example content of just one file. The second file must also exist. I've added example content of `file_2` so it should be clear now

**Comment by julienrf** [comment]
Thank you! That looks good to me.

**Review by alexarchambault — APPROVED** [review]
Looks good, thanks @jkciesluk! (and thanks @julienrf for the review)

**Comment by alexarchambault** [comment]
Oh, I had missed the failing CI job. The error looks legit (looks like a test fixture that needs to be tweaked on Windows)

**Comment by jkciesluk** [comment]
@alexarchambault I tweaked the test, should work now

**Comment by alexarchambault** [comment]
Merging, thanks @jkciesluk!

File: modules/cli/src/main/scala/coursier/cli/params/DependencyParams.scala [hunk 3]
```diff
@@ -61,34 +61,17 @@ object DependencyParams {
...
-          case Left(errors) =>
-            Validated.invalidNel(
-              s"Cannot read dependencies from files:" + System.lineSeparator() +
-                errors
-                  .map("  " + _.getMessage())
-                  .mkString(System.lineSeparator())
-            )
-          case Right(lines) =>
-            Validated.validNel(lines.flatten)
+    val fromFilesDependencies: Seq[String] =
```
Comment (tgodzik) [inline modules/cli/src/main/scala/coursier/cli/params/DependencyParams.scala:65]: @alexarchambault why remove the validation? Wasn't it useful for some better errors? That was done in response to @julienrf 's suggestions

**Pull request merged**